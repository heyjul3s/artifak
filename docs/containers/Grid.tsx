import { Paragraph, Strong } from '../components/Typography';
import { Syntax } from '../components/Syntax';

import {
  columnWidthExample,
  columnWidthMqExample,
  columnLengthExample,
  columnLengthMqExample
} from '../codeExamples/grid/grid';

import {
  ColumnWidthDemo,
  ColumnWidthMqDemo,
  ColumnLengthDemo,
  ColumnLengthMqDemo
} from '../codeExamples/grid/GridDemo';

import { HR } from '../components/HR';
import { FlexTable } from '../components/FlexTable';
import { ArticleDoc } from '../components/ArticleDoc';
import { APIheading } from '../components/APIheading';
import { ArticleSubSectionTitle } from '../components/ArticleSubSectionTitle';

export function Grid() {
  return (
    <ArticleDoc title="Grid">
      <GridContent />
    </ArticleDoc>
  );
}

export function GridContent() {
  return (
    <>
      <Paragraph>
        The Grid library has nothing to with Tron unfortunately and comprises of
        2 components for your layout needs which are the <Strong>Grid</Strong>{' '}
        component and the <Strong>GridItem</Strong> component.
      </Paragraph>

      <HR />

      <APIheading name="Grid" />
      <Paragraph>
        The Grid component comes with a gap default of 1.5em and two params
        namely <strong>columnLength</strong> and <strong>columnWidth</strong>.
      </Paragraph>

      <FlexTable
        title="Props"
        cells={[
          {
            prop: 'columnWidth',
            subProp: 'string | string[]',
            content:
              'This sets a repeating width for the grid auto columns in accordance to the width you provide.'
          },
          {
            prop: 'columnLength',
            subProp: 'number | number[]',
            content:
              ' Similar to columnWidth, except it will be in 1fr and repeats in accordance to the columnLength provided.'
          }
        ]}
      />

      <br />

      <Paragraph>
        That said, you can still set your own gridAutoColumns and style it your
        own way with the Grid component. It will only run the included
        gridAutoColumns styles when columnWidth or columnLength is passed as a
        prop.
      </Paragraph>

      <APIheading name="GridItem" />
      <Paragraph>
        GridItem, as suggested, is a component that is used to contain your
        elements within the Grid component. It includes CSS grid child style
        params to be used at your discretion. By default, it applies no styles
        unless specified.
      </Paragraph>

      <HR />

      <ArticleSubSectionTitle>Usage</ArticleSubSectionTitle>
      <Paragraph>Below are some examples of Grid in action.</Paragraph>

      <APIheading name="columnWidth" />
      <Paragraph>
        With columnWidth, GridItems will wrap around depending on your viewport
        width. You can define a single value...
      </Paragraph>

      <Syntax>{columnWidthExample}</Syntax>

      <ColumnWidthDemo />

      <Paragraph>
        Or for finer control, define an array of values that corresponds to
        breakpoints defined in your theme. Resize your browser window to see it
        in action.
      </Paragraph>

      <Syntax>{columnWidthMqExample}</Syntax>

      <ColumnWidthMqDemo />

      <HR />

      <APIheading name="columnLength" />
      <Paragraph>
        With columnLength, GridItems will retain the specified columnLength and
        resize in accordance to your viewport width.
      </Paragraph>

      <Syntax>{columnLengthExample}</Syntax>

      <ColumnLengthDemo />

      <Paragraph>
        Like its brethren columnWidth, you may also define an array of column
        lengths for finer control. Resize your browser window to see it in
        action.
      </Paragraph>

      <Syntax>{columnLengthMqExample}</Syntax>

      <ColumnLengthMqDemo />
    </>
  );
}
