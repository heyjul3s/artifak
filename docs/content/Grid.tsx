import { H1, H2, Paragraph, LargeLead } from '../components/Typography';
import { Syntax, Params } from '../components/Syntax';
import {
  gridExampleUsage,
  columnWidthExample,
  columnWidthMqExample,
  columnLengthExample,
  columnLengthMqExample
} from '../codeExamples/grid/grid';
import {
  ColumnWidthDemo,
  ColumnWidthMqDemo,
  ColumnLengthDemo,
  ColumnLengthMqDemo
} from '../codeExamples/grid/GridDemo';

export function Grid() {
  return (
    <>
      <H1>Grid</H1>

      <Paragraph>
        The Grid library has nothing to with Tron unfortunately and comprises of
        2 components for your layout needs which are the <strong>Grid</strong>{' '}
        component and the <strong>GridItem</strong> component.
      </Paragraph>

      <LargeLead>Grid</LargeLead>

      <Params
        properties={{
          columnWidth: 'string | string[]',
          columnLength: 'number | number[]'
        }}
      />

      <Paragraph>
        The Grid component comes with a gap default of 1.5em and two properties
        namely <strong>columnLength</strong> and <strong>columnWidth</strong>.
      </Paragraph>

      <LargeLead>GridItem</LargeLead>

      <Paragraph>
        GridItem, as suggested, is a component that is used to contain your
        elements within the Grid component. It includes CSS grid child style
        properties to be used at your discretion. By default, it applies no
        styles unless specified.
      </Paragraph>

      <H2>Usage</H2>

      <LargeLead>columnWidth</LargeLead>

      <Paragraph>
        With columnWidth, GridItems will wrap around depending on your viewport
        width. You can define a single value...
      </Paragraph>

      <Syntax>{columnWidthExample}</Syntax>

      <ColumnWidthDemo />

      <Paragraph>
        Or define an array of values that corresponds to breakpoints defined in
        your theme. Resize your browser window to see it in action.
      </Paragraph>

      <Syntax>{columnWidthMqExample}</Syntax>

      <ColumnWidthMqDemo />

      <LargeLead>columnLength</LargeLead>

      <Paragraph>
        With columnLength, GridItems will retain the specified columnLength and
        resize in accordance to your viewport width.
      </Paragraph>

      <Syntax>{columnLengthExample}</Syntax>

      <ColumnLengthDemo />

      <Paragraph>
        Like its brethren columnWidth, you may also define an array of column
        lengths. Resize your browser window to see it in action.
      </Paragraph>

      <Syntax>{columnLengthMqExample}</Syntax>

      <ColumnLengthMqDemo />
    </>
  );
}
