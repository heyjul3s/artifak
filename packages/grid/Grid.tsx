import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { TGrid, TGridProps } from './typings';

import {
  compose,
  display,
  grid,
  justifyItems,
  justifyContent,
  alignItems,
  alignContent,
  space,
} from 'styled-system';

export const setGridColumnsAndRows = (columnWidth, columnLength) => ({
  ...gridTemplateColumns(columnWidth, columnLength),
});

export const gridTemplateColumns = (
  columnWidth: string,
  columnLength: number
): object => {
  if (columnWidth) {
    return setRepeatableColumnWidth(columnWidth);
  }

  if (columnLength) {
    return setRepeatableColumnLength(columnLength);
  }

  return {};
};

export const setRepeatableColumnWidth = (columnWidth: string): object => ({
  gridTemplateColumns: `repeat(auto-fit, minmax(${columnWidth}, 1fr))`,
});

export const setRepeatableColumnLength = (columnLength: number): object => ({
  gridTemplateColumns: `repeat(${columnLength}, 1fr)`,
});

export const StyledGrid = styled('div')<TGridProps>(
  ({ columnWidth, columnLength, gap = '1.5rem' }) => ({
    display: 'grid',
    gridGap: `${gap}`,
    ...setGridColumnsAndRows(columnWidth, columnLength),
  }),

  compose(
    display,
    grid,
    space,
    alignItems,
    alignContent,
    justifyItems,
    justifyContent
  )
);

export const Grid = forwardRef(
  (
    { columnLength, columnWidth, rowLength, rowHeight, gap, ...props }: TGrid,
    ref
  ) => (
    <StyledGrid
      columnLength={columnLength}
      columnWidth={columnWidth}
      ref={ref}
      {...props}
    />
  )
);
