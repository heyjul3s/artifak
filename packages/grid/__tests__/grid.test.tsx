import React from 'react';
import { setGridTemplateColumns, addGridGap } from '../Grid';
import { Grid, GridItem } from '../';
import { render } from '@testing-library/react';
import 'jest-styled-components';

describe('@artifak/grid', () => {
  describe('setGridTemplateColumns', () => {
    it('should return a string auto-fit CSS grid rule', () => {
      const expected = 'repeat(auto-fit, minmax(320px, 1fr))';
      const gridTemplateColumnStyles = setGridTemplateColumns('320px', void 0);
      expect(gridTemplateColumnStyles).toEqual(expected);
    });

    it('should return an array of auto-fit CSS grid rule', () => {
      const expected = [
        'repeat(auto-fit, minmax(120px, 1fr))',
        'repeat(auto-fit, minmax(320px, 1fr))',
      ];
      const gridTemplateColumnStyles = setGridTemplateColumns(
        ['120px', '320px'],
        void 0
      );
      expect(gridTemplateColumnStyles).toEqual(expected);
    });

    it('should return a string column length CSS grid rule', () => {
      const expected = 'repeat(6, 1fr)';
      expect(setGridTemplateColumns(void 0, 6)).toEqual(expected);
    });

    it('should return a array column length CSS grid rule', () => {
      const expected = ['repeat(2, 1fr)', 'repeat(6, 1fr)'];
      expect(setGridTemplateColumns(void 0, [2, 6])).toEqual(expected);
    });
  });

  describe('addGridGap', () => {
    it('should return string value of "1.5rem" if provided with invalid arguments', () => {
      const expected = '1.5rem';
      expect(addGridGap(void 0)).toEqual(expected);
    });

    it('should return grid gap string value when provided', () => {
      const expected = '3em';
      expect(addGridGap('3em')).toEqual(expected);
    });
  });

  describe('Grid', () => {
    it('should render', () => {
      const { asFragment } = render(
        <Grid>
          <GridItem>1</GridItem>
          <GridItem>2</GridItem>
          <GridItem>3</GridItem>
          <GridItem>4</GridItem>
          <GridItem>5</GridItem>
          <GridItem>6</GridItem>
        </Grid>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('should render grid-auto-columns style if no columnSize prop is provided', () => {
      const { asFragment, getByTestId } = render(
        <Grid data-testid="grid" gridAutoColumns="200px">
          <GridItem>1</GridItem>
          <GridItem>2</GridItem>
          <GridItem>3</GridItem>
          <GridItem>4</GridItem>
          <GridItem>5</GridItem>
          <GridItem>6</GridItem>
        </Grid>
      );

      expect(getByTestId('grid')).toHaveStyleRule('grid-auto-columns', '200px');
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render gridGap style if a value is provided', () => {
      const { asFragment, getByTestId } = render(
        <Grid data-testid="grid" gridGap="20px">
          <GridItem>1</GridItem>
          <GridItem>2</GridItem>
          <GridItem>3</GridItem>
          <GridItem>4</GridItem>
          <GridItem>5</GridItem>
          <GridItem>6</GridItem>
        </Grid>
      );

      expect(getByTestId('grid')).toHaveStyleRule('grid-gap', '20px');
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
