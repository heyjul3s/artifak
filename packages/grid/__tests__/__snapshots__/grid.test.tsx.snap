// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@artifak/grid Grid should render 1`] = `
<DocumentFragment>
  .c0 {
  display: grid;
  grid-gap: 1.5rem;
}

<div
    class="c0"
    display="grid"
  >
    <div
      class=""
    >
      1
    </div>
    <div
      class=""
    >
      2
    </div>
    <div
      class=""
    >
      3
    </div>
    <div
      class=""
    >
      4
    </div>
    <div
      class=""
    >
      5
    </div>
    <div
      class=""
    >
      6
    </div>
  </div>
</DocumentFragment>
`;

exports[`@artifak/grid Grid should render grid-auto-columns style if no columnSize prop is provided 1`] = `
<DocumentFragment>
  .c0 {
  display: grid;
  grid-gap: 1.5rem;
  grid-auto-columns: 200px;
}

<div
    class="c0"
    data-testid="grid"
    display="grid"
  >
    <div
      class=""
    >
      1
    </div>
    <div
      class=""
    >
      2
    </div>
    <div
      class=""
    >
      3
    </div>
    <div
      class=""
    >
      4
    </div>
    <div
      class=""
    >
      5
    </div>
    <div
      class=""
    >
      6
    </div>
  </div>
</DocumentFragment>
`;

exports[`@artifak/grid Grid should render gridGap style if a value is provided 1`] = `
<DocumentFragment>
  .c0 {
  display: grid;
  grid-gap: 20px;
}

<div
    class="c0"
    data-testid="grid"
    display="grid"
  >
    <div
      class=""
    >
      1
    </div>
    <div
      class=""
    >
      2
    </div>
    <div
      class=""
    >
      3
    </div>
    <div
      class=""
    >
      4
    </div>
    <div
      class=""
    >
      5
    </div>
    <div
      class=""
    >
      6
    </div>
  </div>
</DocumentFragment>
`;
